src/stan/variational/advi.hpp

```diff
@@ -0,0 +1,665 @@
+           std::ostream* diagnostic_stream):
+        model_(m),
+        cont_params_(cont_params),
+        elbo_(elbo),
+        rng_(rng),
+        n_monte_carlo_grad_(n_monte_carlo_grad),
+        n_monte_carlo_elbo_(n_monte_carlo_elbo),
+        eta_stepsize_(eta_stepsize),
+        refresh_(refresh),
+        print_stream_(print_stream),
+        out_stream_(output_stream),
+        diag_stream_(diagnostic_stream) {
+        static const char* function =
+          "stan::variational::advi";
+
+        stan::math::check_positive(function,
+                                 "Number of Monte Carlo samples for gradients",
+                                 n_monte_carlo_grad_);
+
+        stan::math::check_positive(function,
+                                   "Number of Monte Carlo samples for ELBO",
+                                   n_monte_carlo_elbo_);
+
+        stan::math::check_positive(function, "Eta stepsize", eta_stepsize_);
+        }
+
+      virtual ~advi() {}
+
+
+      /**
+       * ELBO
+       *
+       * Calculates the "blackbox" Evidence Lower BOund (ELBO) by sampling
+       * from the standard multivariate normal (for now), affine transform
+       * the sample, and evaluating the log joint, adjusted by the entropy
+       * term of the normal
+       *
+       * @param   advi_params      variational parameters class
+       * @return                   evidence lower bound (elbo)
+       */
+      template <typename T>
+      double calc_ELBO(const T& advi_params) {
+        double elbo(0.0);
+        int dim = advi_params.dimension();
+
+        Eigen::VectorXd z_check   = Eigen::VectorXd::Zero(dim);
+        Eigen::VectorXd z_tilde   = Eigen::VectorXd::Zero(dim);
+
+        for (int i = 0; i < n_monte_carlo_elbo_; ++i) {
+          // Draw from standard normal and transform to unconstrained space
+          for (int d = 0; d < dim; ++d) {
+            z_check(d) = stan::prob::normal_rng(0, 1, rng_);
+          }
+          z_tilde = advi_params.to_unconstrained(z_check);
+
+          // Accumulate log probability
+          elbo += (model_.template
+                   log_prob<false, true>(z_tilde, print_stream_));
+        }
+        // Divide to get Monte Carlo integral estimate
+        elbo /= static_cast<double>(n_monte_carlo_elbo_);
+
+        // Add entropy term
+        elbo += advi_params.entropy();
+
+        return elbo;
+      }
+
+      /**
+       * FULL-RANK GRADIENTS
+       *
+       * Calculates the "blackbox" gradient with respect to BOTH the location
+       * vector (mu) and the cholesky factor of the scale matrix (L) in
+       * parallel. It uses the same gradient computed from a set of Monte Carlo
+       * samples
+       *
+       * @param muL     mean and cholesky factor of affine transform
+       * @param mu_grad gradient of location vector parameter
+       * @param L_grad  gradient of scale matrix parameter
+       */
+      void calc_combined_grad(const advi_params_fullrank& muL,
```

Comment:
Tastes differ on whether to prefer standalone functions
or members.  I prefer members, but standalone's not so bad.

But that comment "FULL-RANK GRADIENTS" is bad --- the first
sentence should succinctly describe what the function takes
as inputs and what it returns. 

- Bob

> On Apr 29, 2015, at 3:43 AM, Rob Trangucci <notifications@github.com> wrote:
> 
> In src/stan/variational/advi.hpp:
> 
> > +        return elbo;
> > +      }
> > +
> > +      /**
> > +       * FULL-RANK GRADIENTS
> > +       *
> > +       * Calculates the "blackbox" gradient with respect to BOTH the location
> > +       * vector (mu) and the cholesky factor of the scale matrix (L) in
> > +       * parallel. It uses the same gradient computed from a set of Monte Carlo
> > +       * samples
> > +       *
> > +       * @param muL     mean and cholesky factor of affine transform
> > +       * @param mu_grad gradient of location vector parameter
> > +       * @param L_grad  gradient of scale matrix parameter
> > +       */
> > +      void calc_combined_grad(const advi_params_fullrank& muL,
> 
> @akucukelbir Should be a member function of advi_paramas_fullrank, doesn't make sense as a standalone.
> 
> —
> Reply to this email directly or view it on GitHub.
>